pkg/test/framework/components/echo/echoboot/echoboot.go

```diff
@@ -43,42 +75,166 @@ func (b builder) WithConfig(cfg echo.Config) echo.Builder {
 // to that cluster, otherwise the Config is applied to all WithClusters. Once built, if being built for a sngle cluster,
 // the instance pointer will be updated to point at the new Instance.
 func (b builder) With(i *echo.Instance, cfg echo.Config) echo.Builder {
-	// TODO support for other kinds of cluster
+	cfg = cfg.DeepCopy()
+	common.FillInDefaults(&cfg)
 
-	// Per-config cluster override, or no WithClusters provided
-	if cfg.Cluster != nil || len(b.clusters) == 0 {
-		b.kubeBuilder = b.kubeBuilder.With(i, cfg)
-		return b
+	targetClusters := b.clusters
+	if cfg.Cluster != nil {
+		targetClusters = cluster.Clusters{cfg.Cluster}
 	}
 
-	// WithClusters
-	// only provide the ref for a single cluster
-	var ref *echo.Instance
-	if len(b.clusters) == 1 {
-		ref = i
+	deployedTo := 0
+	for idx, c := range targetClusters {
+		ec, ok := c.(echo.Cluster)
+		if !ok {
+			b.errs = multierror.Append(b.errs, fmt.Errorf("attembed to deploy to %s but it does not implement echo.Cluster", c.Name()))
+			continue
+		}
+		if perClusterConfig, ok := ec.CanDeploy(cfg); ok {
+			var ref *echo.Instance
+			if idx == 0 {
+				// ref only applies to the first cluster deployed to
+				// refs shouldn't be used when deploying to multiple targetClusters
+				// TODO: should we just panic if a ref is passed in a multi-cluster context?
+				ref = i
+			}
+			perClusterConfig = perClusterConfig.DeepCopy()
+			k := ec.Kind()
+			perClusterConfig.Cluster = ec
+			b.configs[k] = append(b.configs[k], perClusterConfig)
+			b.refs[k] = append(b.refs[k], ref)
+			deployedTo++
+		}
 	}
-	// provide a config for each cluster to the kubeBuilder
-	for _, cluster := range b.clusters {
-		// TODO add a mechanism to allow "claiming" a config, and not providing it to other clusters
-		clusterCfg := cfg
-		clusterCfg.Cluster = cluster
-		b.kubeBuilder.With(ref, clusterCfg)
+
+	// If we didn't deploy VMs, but we don't care about VMs, we can ignore this.
+	shouldSkip := b.ctx.Settings().SkipVM && cfg.DeployAsVM
```

Comment:
Do we eventually want to get rid of `DeployAsVM`? It still seems that this means the test has to think about VMs specifically, which means low test coverage.

Reply:
I think we really need it inside of echo. Tests shouldn't think about deploying apps, suites should. And suites shouldn't really do too much customization, they should just get a common set of apps (pilot uses `common/apps.go`).

When we get our standard set of apps we have to enumerate what's special about each app (VM, headless, naked, external, etc.), so it's good to specify this. Most new tests won't have to think about _including_ vms if they're using a common "echo set". 

The tests should stop defining a list of clients/targets. What we have now _does_ make the tests enumerate these things. Should just be a double loop over `apps.All`:

https://github.com/istio/istio/blob/2fd4dfd61713e65b1974ca16257ad21d8205b584/tests/integration/pilot/common/routing.go#L802-L808

Tests and suites alike though, we should let them skip a test or remove the vm from the deployments if it has some incompatibility with VMs like

```go
apps.EnumerateCases("sniffing", func (ctx framework.TestContext src echo.Instance, dst echo.Instances) {
  if (app.Config().DeployAsVM) {
    t.skip("github.com/issue/123 - VMs are broken for my feature")
  }
})
```